### PR TITLE
feat: send and receive unflattened public inputs to backend

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/3547

## Summary\*

This PR changes the public inputs being passed to the backend to being a `WitnessMap` rather than an array of `Uint8Array`s. This is very useful for us to be able to pull public input values from the object returned from the backend as we can just abi decode it with the public abi.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/noir_js_backend_barretenberg/src/index.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ import { decompressSync as gunzip } from 'fflate';
 import { acirToUint8Array } from './serialize.js';
 import { Backend, CompiledCircuit, ProofData } from '@noir-lang/types';
 import { BackendOptions } from './types.js';
+import { WitnessMap } from '@noir-lang/noirc_abi';
```
Comment (TomAFrench) [inline tooling/noir_js_backend_barretenberg/src/index.ts:6]: Expose through `@noir-lang/types` now that it's part of the backend interface.

File: tooling/noir_js_backend_barretenberg/src/index.ts [hunk 2]
```diff
@@ -93,13 +94,32 @@ export class BarretenbergBackend implements Backend {
...
+    const abi = this.acirCircuit.abi;
+    const return_value_witnesses = abi.return_witnesses;
+    const public_parameters = abi.parameters.filter((param) => param.visibility === 'public');
+    const public_parameter_witnesses: number[] = public_parameters.flatMap((param) =>
+      abi.param_witnesses[param.name].flatMap((witness_range) =>
+        Array.from({ length: witness_range.end - witness_range.start }, (_, i) => witness_range.start + i),
+      ),
+    );
+
+    // We now have an array of witness indices which have been duplicated and sorted in ascending order.
```
Comment (TomAFrench) [inline tooling/noir_js_backend_barretenberg/src/index.ts:113]:
```suggestion
    // We now have an array of witness indices which have been deduplicated and sorted in ascending order.
```

File: compiler/integration-tests/test/node/smart_contract_verifier.test.ts
Comment (TomAFrench) [inline compiler/integration-tests/test/node/smart_contract_verifier.test.ts:1]: Bit sad how this is leaking into here. We should have some helper functions to switch between witnessmap and flattened array but we need to decide on the best place to put it.

I think `@noir-lang/noir_js_barretenberg_backend` would be best as the flattened format is an implementation detail of bberg. There's no guarantee that a smart contract implementation from another backend would use this same format.

File: tooling/noir_js_backend_barretenberg/src/index.ts [hunk 3]
```diff
@@ -91,16 +94,8 @@ export class BarretenbergBackend implements Backend {
...
-    const publicInputSize = 32;
-    const publicInputs: Uint8Array[] = [];
-
-    for (let i = 0; i < publicInputsConcatenated.length; i += publicInputSize) {
-      const publicInput = publicInputsConcatenated.slice(i, i + publicInputSize);
-      publicInputs.push(publicInput);
-    }
-
     const proof = proofWithPublicInputs.slice(splitIndex);
+    const publicInputs = deflattenPublicInputs(publicInputsConcatenated, this.acirCircuit.abi);
```
Comment (TomAFrench) [inline tooling/noir_js_backend_barretenberg/src/index.ts:98]: This is slightly bad as it requires a "backend" to know about the circuit ABI. We have a path to remove this however as the public inputs are also detailed within the "bytecode" object (as this is a serialized `Circuit` struct).

Comment (kevaundray) [reply tooling/noir_js_backend_barretenberg/src/index.ts:98]: can you create an issue for this?

**Pull request merged**